R/cal-estimate.R

```diff
@@ -0,0 +1,405 @@
+#' # dplyr selector functions are also supported
+#' cal_estimate_isotonic(segment_logistic, Class, dplyr::starts_with(".pred_"))
+#' @export
+cal_estimate_isotonic <- function(.data,
+                                  truth = NULL,
+                                  estimate = dplyr::starts_with(".pred_"),
+                                  ...) {
+  UseMethod("cal_estimate_isotonic")
+}
+
+#' @export
+cal_estimate_isotonic.data.frame <- function(.data,
+                                             truth = NULL,
+                                             estimate = dplyr::starts_with(".pred_"),
+                                             ...) {
+  truth <- enquo(truth)
+
+  levels <- truth_estimate_map(.data, {{ truth }}, {{ estimate }})
+
+  if (length(levels) == 2) {
+    iso_model <- cal_isoreg_dataframe(
+      .data = .data,
+      truth = !!truth,
+      estimate = levels[[1]],
+      ...
+    )
+
+    res <- as_binary_cal_object(
+      estimate = iso_model,
+      levels = levels,
+      truth = !!truth,
+      method = "Isotonic",
+      additional_class = "cal_estimate_isotonic"
+    )
+  } else {
+    stop_multiclass()
+  }
+
+  res
+}
+
+#------------------------ >> Isotonic Bootstrapped------------------------------
+#' Uses a bootstrapped Isotonic regression model to calibrate probabilities
+#' @param times Number of bootstraps.
+#' @inheritParams cal_estimate_logistic
+#' @examples
+#' # It will automatically identify the probability columns
+#' # if passed a model fitted with tidymodels
+#' cal_estimate_isotonic_boot(segment_logistic, Class)
+#' # Specify the variable names in a vector of unquoted names
+#' cal_estimate_isotonic_boot(segment_logistic, Class, c(.pred_poor, .pred_good))
+#' # dplyr selector functions are also supported
+#' cal_estimate_isotonic_boot(segment_logistic, Class, dplyr::starts_with(".pred_"))
+#' @export
+cal_estimate_isotonic_boot <- function(.data,
+                                       truth = NULL,
+                                       estimate = dplyr::starts_with(".pred_"),
+                                       times = 10,
+                                       ...) {
+  UseMethod("cal_estimate_isotonic_boot")
+}
+
+#' @export
+cal_estimate_isotonic_boot.data.frame <- function(.data,
+                                                  truth = NULL,
+                                                  estimate = dplyr::starts_with(".pred_"),
+                                                  times = 10,
+                                                  ...) {
+  truth <- enquo(truth)
+
+  levels <- truth_estimate_map(.data, !!truth, {{ estimate }})
+
+  if (length(levels) == 2) {
+    log_model <- cal_isoreg_boot(
+      .data = .data,
+      truth = !!truth,
+      estimate = levels[[1]],
+      times = times,
+      ...
+    )
+
+    res <- as_binary_cal_object(
+      estimate = log_model,
+      levels = levels,
+      truth = !!truth,
+      method = "Isotonic Bootstrapped",
```

Comment:
We should use methods of "Isotonic Regression" and "Bootstrapped Isotonic Regression"